src/backend/joanie/payment/models.py

```diff
@@ -1 +1,463 @@
+                        "product": {
+                            "name": related_product.title,  # pylint: disable=no-member
+                            "description": related_product.description,  # noqa pylint: disable=no-member,line-too-long
+                        }
+                    }
+                }
+
+        self.localized_context = context
+
+    def get_document_context(self, language_code=None):
+        """
+        Build the invoice document context for the given language.
+        If no language_code is provided, we use the active language.
+        """
+        language_code = language_code or get_language()
+
+        vat = D(settings.JOANIE_VAT)
+        total_amount = self.amount.amount  # pylint: disable=no-member
+        vat_amount = total_amount * vat / 100
+        net_amount = total_amount - vat_amount
+        currency = get_currency_symbol(
+            self.amount.currency.code  # pylint: disable=no-member
+        )
+
+        base_context = {
+            "metadata": {
+                "issued_on": str(self.updated_on),
+                "reference": self.reference,
+                "type": self.type,
+            },
+            "order": {
+                "amount": {
+                    "currency": currency,
+                    "subtotal": str(net_amount),
+                    "total": str(total_amount),
+                    "vat_amount": str(vat_amount),
+                    "vat": str(vat),
+                },
+                "company": settings.JOANIE_INVOICE_COMPANY_CONTEXT,
+                "customer": {
+                    "address": self.recipient_address,
+                    "name": self.recipient_name,
+                },
+                "seller": {
+                    "address": settings.JOANIE_INVOICE_SELLER_ADDRESS,
+                },
+            },
+        }
+
+        try:
+            localized_context = self.localized_context[language_code]
+        except KeyError:
+            fallback_language = settings.PARLER_DEFAULT_LANGUAGE_CODE
+            localized_context = self.localized_context[fallback_language]
+
+        return merge_dict(base_context, localized_context)
+
+    def normalize_reference(self):
+        """
+        Generate a normalized reference related to the date
+        and the related order
+        """
+        order_uid_fragment = str(self.get_related_order().uid).split("-", maxsplit=1)[0]
+        timestamp = int(timezone.now().timestamp() * 1_000)  # Time in milliseconds
+
+        return f"{order_uid_fragment}-{timestamp}"
+
+    def clean(self):
+        if (self.order is None and self.parent is None) or (self.order and self.parent):
+            raise ValidationError(_("Invoice must have an order or a parent."))
+
+        if self.parent:
+            self.recipient_name = self.parent.recipient_name
+            self.recipient_address = self.parent.recipient_address
+
+        if self.type == payment_enums.INVOICE_TYPE_CREDIT_NOTE:
```

Comment:
should we also check (maybe with a db constraint more than here in the clean?) that an invoice with a parent has a negative amount? and an invoice with an order has a positive amount?

Reply:
Nope because credit note can have sibling invoice.